storage/innobase/include/trx0purge.h

```diff
@@ -243,20 +238,29 @@ class purge_sys_t
 					record */
 	uint16_t	hdr_offset;	/*!< Header byte offset on the page */
 
-  /** Binary min-heap of indexes in trx_sys.rseg_array, ordered on
-  rseg_t::last_trx_no(). It is protected by the pq_mutex */
+  /** Binary min-heap of (trx_no, trx_sys.rseg_array index) pairs, ordered on
+  trx_no. It is protected by the pq_mutex */
   purge_queue purge_queue;
 
   /** Mutex protecting purge_queue */
   mysql_mutex_t pq_mutex;
 
 public:
+
+  void enqueue(trx_id_t trx_no, const trx_rseg_t &rseg) {
+    mysql_mutex_assert_owner(&pq_mutex);
+    purge_queue.push(trx_no, rseg);
+  }
+
   /** Push to purge queue without acquiring pq_mutex.
   @param rseg rseg to push */
-  void enqueue(trx_rseg_t &rseg)
-  {
+  void enqueue(const trx_rseg_t &rseg) { enqueue(rseg.last_trx_no(), rseg); }
+
+  /** Clone purge queue container.
+  @return purge queue container clone */
+  purge_queue::container_type clone_queue_container() const {
     mysql_mutex_assert_owner(&pq_mutex);
-    purge_queue.push(&rseg);
+    return purge_queue.clone_container();
   }
```

Comment:
Based on the following test program that I checked on https://godbolt.org, this second level of copy propagation will indeed be elided starting with GCC 4.8.5 `-O1`:
```c++
#include <vector>
typedef std::vector<unsigned> vect;
static vect clone(const vect &v) { return v; }
static vect klone(const vect &v) { return clone(v); }
bool f(const vect &v)
{
  vect m = klone(v);
  return m.empty();
}
```
If I remove the `static` (which encourages the compiler to `inline` the calls), then the function `klone()` does not reduce to a simple tail call to `clone()`, not even when using the most recent compilers (GCC 13 or clang 18).

I compiled cb08831e7249039b279c08f819efe3d085d32523 using GCC 13.2.0 with both `-Og` (debug build) and `-O2`, and the extra copies are being elided just fine.